File: labs/03-harness/harness-solution/product-inventory/ContosoInventory.Server/Services/ProductService.cs
using Microsoft.EntityFrameworkCore;
using ContosoInventory.Server.Data;
using ContosoInventory.Server.Models;
using ContosoInventory.Shared.DTOs;

namespace ContosoInventory.Server.Services;

/// <summary>
/// Provides operations for managing inventory products.
/// </summary>
public class ProductService : IProductService
{
    private readonly InventoryContext _context;
    private readonly ILogger<ProductService> _logger;

    public ProductService(InventoryContext context, ILogger<ProductService> logger)
    {
        _context = context;
        _logger = logger;
    }

    /// <inheritdoc />
    public async Task<List<ProductResponseDto>> GetAllProductsAsync(int? categoryId)
    {
        try
        {
            var query = _context.Products
                .AsNoTracking()
                .Include(p => p.Category)
                .AsQueryable();

            if (categoryId.HasValue)
            {
                query = query.Where(p => p.CategoryId == categoryId.Value);
            }

            var products = await query
                .OrderBy(p => p.Name)
                .ToListAsync();

            return products.Select(MapToDto).ToList();
        }
        catch (Exception ex)
        {
            _logger.LogError(ex, "Error retrieving all products.");
            throw;
        }
    }

    /// <inheritdoc />
    public async Task<ProductResponseDto?> GetProductByIdAsync(int id)
    {
        try
        {
            var product = await _context.Products
                .AsNoTracking()
                .Include(p => p.Category)
                .FirstOrDefaultAsync(p => p.Id == id);

            return product == null ? null : MapToDto(product);
        }
        catch (Exception ex)
        {
            _logger.LogError(ex, "Error retrieving product with ID {ProductId}.", id);
            throw;
        }
    }

    /// <inheritdoc />
    public async Task<ProductResponseDto> CreateProductAsync(CreateProductDto dto)
    {
        try
        {
            var categoryExists = await _context.Categories.AnyAsync(c => c.Id == dto.CategoryId);
            if (!categoryExists)
            {
                throw new InvalidOperationException($"Category with ID {dto.CategoryId} doesn't exist.");
            }

            var skuExists = await _context.Products.AnyAsync(p => p.Sku.ToLower() == dto.Sku.ToLower());
            if (skuExists)
            {
                throw new InvalidOperationException($"A product with the SKU '{dto.Sku}' already exists.");
            }

            var product = new Product
            {
                Name = dto.Name,
                Sku = dto.Sku,
                Description = dto.Description,
                Price = dto.Price,
                StockQuantity = dto.StockQuantity,
                CategoryId = dto.CategoryId,
                CreatedDate = DateTime.UtcNow,
                LastUpdatedDate = DateTime.UtcNow
            };

            _context.Products.Add(product);
            await _context.SaveChangesAsync();

            _logger.LogInformation("Product created: {ProductName} (ID: {ProductId}).", product.Name, product.Id);

            return await GetProductByIdAsync(product.Id)
                ?? throw new InvalidOperationException("Failed to retrieve the product after creation.");
        }
        catch (InvalidOperationException)
        {
            throw;
        }
        catch (Exception ex)
        {
            _logger.LogError(ex, "Error creating product '{ProductName}'.", dto.Name);
            throw;
        }
    }

    /// <inheritdoc />
    public async Task<ProductResponseDto?> UpdateProductAsync(int id, UpdateProductDto dto)
    {
        try
        {
            var product = await _context.Products.FindAsync(id);
            if (product == null)
            {
                return null;
            }

            var categoryExists = await _context.Categories.AnyAsync(c => c.Id == dto.CategoryId);
            if (!categoryExists)
            {
                throw new InvalidOperationException($"Category with ID {dto.CategoryId} doesn't exist.");
            }

            var skuExists = await _context.Products
                .AnyAsync(p => p.Sku.ToLower() == dto.Sku.ToLower() && p.Id != id);
            if (skuExists)
            {
                throw new InvalidOperationException($"A product with the SKU '{dto.Sku}' already exists.");
            }

            product.Name = dto.Name;
            product.Sku = dto.Sku;
            product.Description = dto.Description;
            product.Price = dto.Price;
            product.StockQuantity = dto.StockQuantity;
            product.CategoryId = dto.CategoryId;
            product.LastUpdatedDate = DateTime.UtcNow;

            await _context.SaveChangesAsync();

            _logger.LogInformation("Product updated: {ProductName} (ID: {ProductId}).", product.Name, product.Id);

            return await GetProductByIdAsync(product.Id);
        }
        catch (InvalidOperationException)
        {
            throw;
        }
        catch (Exception ex)
        {
            _logger.LogError(ex, "Error updating product with ID {ProductId}.", id);
            throw;
        }
    }

    /// <inheritdoc />
    public async Task<bool> DeleteProductAsync(int id)
    {
        try
        {
            var product = await _context.Products.FindAsync(id);
            if (product == null)
            {
                return false;
            }

            _context.Products.Remove(product);
            await _context.SaveChangesAsync();

            _logger.LogInformation("Product deleted: {ProductName} (ID: {ProductId}).", product.Name, product.Id);

            return true;
        }
        catch (Exception ex)
        {
            _logger.LogError(ex, "Error deleting product with ID {ProductId}.", id);
            throw;
        }
    }

    /// <inheritdoc />
    public async Task<ProductResponseDto?> RestockAsync(int id, int quantity)
    {
        try
        {
            var product = await _context.Products.FindAsync(id);
            if (product == null)
            {
                return null;
            }

            product.StockQuantity += quantity;
            product.LastUpdatedDate = DateTime.UtcNow;

            await _context.SaveChangesAsync();

            _logger.LogInformation("Product restocked: {ProductName} (ID: {ProductId}) by {Quantity}. New stock: {StockQuantity}.",
                product.Name, product.Id, quantity, product.StockQuantity);

            return await GetProductByIdAsync(product.Id);
        }
        catch (Exception ex)
        {
            _logger.LogError(ex, "Error restocking product with ID {ProductId}.", id);
            throw;
        }
    }

    private static ProductResponseDto MapToDto(Product product)
    {
        return new ProductResponseDto
        {
            Id = product.Id,
            Name = product.Name,
            Sku = product.Sku,
            Description = product.Description,
            Price = product.Price,
            StockQuantity = product.StockQuantity,
            CategoryId = product.CategoryId,
            CategoryName = product.Category?.Name ?? string.Empty,
            CreatedDate = product.CreatedDate,
            LastUpdatedDate = product.LastUpdatedDate
        };
    }
}

File: labs/03-harness/harness-solution/product-inventory-py/app/services/product_service.py
"""Operations for managing inventory products."""

import logging
import sqlite3
from datetime import datetime, timezone
from typing import Protocol

from app.schemas.product import CreateProductDto, ProductResponseDto, UpdateProductDto

logger = logging.getLogger(__name__)

_SELECT_PRODUCT = (
    'SELECT p."Id", p."Name", p."Sku", p."Description", p."Price", p."StockQuantity", '
    'p."CategoryId", c."Name" AS "CategoryName", p."CreatedDate", p."LastUpdatedDate" '
    'FROM "Products" p JOIN "Categories" c ON c."Id" = p."CategoryId"'
)


class ProductServiceProtocol(Protocol):
    """Operations for managing inventory products."""

    def get_all_products(self, category_id: int | None) -> list[ProductResponseDto]:
        """Return all products, optionally filtered by category."""

    def get_product_by_id(self, product_id: int) -> ProductResponseDto | None:
        """Return a product, or None if it does not exist."""

    def create_product(self, dto: CreateProductDto) -> ProductResponseDto:
        """Create a product. Raises ValueError for a duplicate SKU or unknown category."""

    def update_product(self, product_id: int, dto: UpdateProductDto) -> ProductResponseDto | None:
        """Update a product, or return None if it does not exist. Raises ValueError like create."""

    def delete_product(self, product_id: int) -> bool:
        """Delete a product and return whether it existed."""

    def restock(self, product_id: int, quantity: int) -> ProductResponseDto | None:
        """Add stock to a product, or return None if it does not exist."""


class ProductService:
    """SQLite implementation of the product operations."""

    def __init__(self, connection: sqlite3.Connection) -> None:
        self._connection = connection

    def get_all_products(self, category_id: int | None) -> list[ProductResponseDto]:
        """Return all products ordered by name, optionally filtered by category."""
        if category_id is None:
            rows = self._connection.execute(f'{_SELECT_PRODUCT} ORDER BY p."Name"').fetchall()
        else:
            rows = self._connection.execute(
                f'{_SELECT_PRODUCT} WHERE p."CategoryId" = ? ORDER BY p."Name"', (category_id,)
            ).fetchall()
        return [_map_to_dto(row) for row in rows]

    def get_product_by_id(self, product_id: int) -> ProductResponseDto | None:
        """Return a product, or None if it does not exist."""
        row = self._connection.execute(f'{_SELECT_PRODUCT} WHERE p."Id" = ?', (product_id,)).fetchone()
        return None if row is None else _map_to_dto(row)

    def create_product(self, dto: CreateProductDto) -> ProductResponseDto:
        """Create a product. Raises ValueError for a duplicate SKU or unknown category."""
        self._ensure_category_exists(dto.category_id)
        self._ensure_sku_is_free(dto.sku, None)
        now = _utc_now()
        with self._connection:
            cursor = self._connection.execute(
                'INSERT INTO "Products" ("Name", "Sku", "Description", "Price", "StockQuantity", '
                '"CategoryId", "CreatedDate", "LastUpdatedDate") VALUES (?, ?, ?, ?, ?, ?, ?, ?)',
                (dto.name, dto.sku, dto.description, dto.price, dto.stock_quantity, dto.category_id, now, now),
            )
        logger.info("Product created: %s (ID: %s).", dto.name, cursor.lastrowid)
        return self.get_product_by_id(cursor.lastrowid)

    def update_product(self, product_id: int, dto: UpdateProductDto) -> ProductResponseDto | None:
        """Update a product, or return None if it does not exist. Raises ValueError like create."""
        if self.get_product_by_id(product_id) is None:
            return None
        self._ensure_category_exists(dto.category_id)
        self._ensure_sku_is_free(dto.sku, product_id)
        with self._connection:
            self._connection.execute(
                'UPDATE "Products" SET "Name" = ?, "Sku" = ?, "Description" = ?, "Price" = ?, '
                '"StockQuantity" = ?, "CategoryId" = ?, "LastUpdatedDate" = ? WHERE "Id" = ?',
                (dto.name, dto.sku, dto.description, dto.price, dto.stock_quantity, dto.category_id,
                 _utc_now(), product_id),
            )
        logger.info("Product updated: %s (ID: %s).", dto.name, product_id)
        return self.get_product_by_id(product_id)

    def delete_product(self, product_id: int) -> bool:
        """Delete a product and return whether it existed."""
        with self._connection:
            cursor = self._connection.execute('DELETE FROM "Products" WHERE "Id" = ?', (product_id,))
        if cursor.rowcount == 0:
            return False
        logger.info("Product deleted: ID %s.", product_id)
        return True

    def restock(self, product_id: int, quantity: int) -> ProductResponseDto | None:
        """Add stock to a product, or return None if it does not exist."""
        with self._connection:
            cursor = self._connection.execute(
                'UPDATE "Products" SET "StockQuantity" = "StockQuantity" + ?, "LastUpdatedDate" = ? WHERE "Id" = ?',
                (quantity, _utc_now(), product_id),
            )
        if cursor.rowcount == 0:
            return None
        product = self.get_product_by_id(product_id)
        logger.info("Product restocked: ID %s by %s. New stock: %s.", product_id, quantity, product.stock_quantity)
        return product

    def _ensure_category_exists(self, category_id: int) -> None:
        row = self._connection.execute('SELECT 1 FROM "Categories" WHERE "Id" = ?', (category_id,)).fetchone()
        if row is None:
            raise ValueError(f"Category with ID {category_id} doesn't exist.")

    def _ensure_sku_is_free(self, sku: str, product_id: int | None) -> None:
        row = self._connection.execute(
            'SELECT 1 FROM "Products" WHERE "Sku" = ? COLLATE NOCASE AND "Id" IS NOT ?', (sku, product_id)
        ).fetchone()
        if row is not None:
            raise ValueError(f"A product with the SKU '{sku}' already exists.")


def _utc_now() -> str:
    return datetime.now(timezone.utc).replace(tzinfo=None).isoformat()


def _map_to_dto(row: sqlite3.Row) -> ProductResponseDto:
    return ProductResponseDto(
        id=row["Id"],
        name=row["Name"],
        sku=row["Sku"],
        description=row["Description"],
        price=row["Price"],
        stock_quantity=row["StockQuantity"],
        category_id=row["CategoryId"],
        category_name=row["CategoryName"],
        created_date=row["CreatedDate"],
        last_updated_date=row["LastUpdatedDate"],
    )
